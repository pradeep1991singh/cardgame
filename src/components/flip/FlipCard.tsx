import React, {FC, memo, useCallback, useEffect, useRef} from 'react';
import {Text, StyleSheet, Animated, Pressable} from 'react-native';
import {useDispatch} from 'react-redux';
import {incrementCardClickCount, validateCard} from '../../store/cardReducer';
import {FlipCardProps} from './types';

const styles = StyleSheet.create({
  flipCard: {
    width: 100,
    height: 150,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#1ca2f4',
    backfaceVisibility: 'hidden',
    borderRadius: 5,
    borderWidth: 2,
    borderColor: '#ffffff',
    margin: 5,
  },
  flipBackCard: {
    backgroundColor: '#ffffff',
    position: 'absolute',
    top: 0,
  },
  text: {
    lineHeight: 25,
    fontSize: 18,
    marginVertical: 18,
    textAlign: 'center',
  },
  textWhite: {
    color: '#ffffff',
  },
});

const FlipCard: FC<FlipCardProps> = memo(({item}) => {
  const {value, index, isOpen} = item;
  const dispatch = useDispatch();
  const animatedValue = useRef(new Animated.Value(0)).current;

  let flipValue = 0;
  animatedValue.addListener(({value: val}) => (flipValue = val));

  const frontAnimation = animatedValue.interpolate({
    inputRange: [0, 180],
    outputRange: ['0deg', '180deg'],
  });
  const backAnimation = animatedValue.interpolate({
    inputRange: [0, 180],
    outputRange: ['180deg', '360deg'],
  });

  const frontAnimationStyle = {
    transform: [{rotateY: frontAnimation}],
  };
  const backAnimationStyle = {
    transform: [{rotateY: backAnimation}],
  };

  const closeCard = useCallback(() => {
    Animated.spring(animatedValue, {
      toValue: 0,
      friction: 8,
      tension: 10,
      useNativeDriver: true,
    }).start();
  }, [animatedValue]);

  const openCard = useCallback(() => {
    Animated.spring(animatedValue, {
      toValue: 180,
      friction: 8,
      tension: 10,
      useNativeDriver: true,
    }).start();
  }, [animatedValue]);

  useEffect(() => {
    if (value) {
      closeCard();
    }
  }, [value, closeCard]);

  useEffect(() => {
    if (!isOpen) {
      closeCard();
    }
  }, [isOpen, closeCard]);

  const flipCardHandler = () => {
    if (flipValue >= 90) {
      closeCard();
    } else {
      openCard();
    }
    dispatch(incrementCardClickCount({value, index, isOpen}));
    setTimeout(() => dispatch(validateCard({value, index, isOpen})), 1000);
  };

  return (
    <Pressable onPress={flipCardHandler}>
      <Animated.View style={[styles.flipCard, frontAnimationStyle]}>
        <Text style={[styles.text, styles.textWhite]}>?</Text>
      </Animated.View>
      <Animated.View
        style={[backAnimationStyle, styles.flipCard, styles.flipBackCard]}>
        <Text style={styles.text}>{value}</Text>
      </Animated.View>
    </Pressable>
  );
});

export default FlipCard;
