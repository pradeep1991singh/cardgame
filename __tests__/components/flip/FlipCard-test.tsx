// @ts
import 'react-native';
import React from 'react';
import {create} from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import FlipCard from '../../../src/components/flip/FlipCard';
import {MockCard} from '../../../__mocks__/card.mock';
import {Provider} from 'react-redux';
import {Pressable} from 'react-native';

const mockStore = configureStore([]);

describe('FlipCard', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = jest.fn();
    component = create(
      <Provider store={store}>
        <FlipCard item={MockCard} />
      </Provider>,
    );
  });

  it('renders correctly && snapshot', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('flipCardHandler', () => {
    expect(component.root.instance.refs).toBe(1);
    const flipCard = component.root.findByType(Pressable);
    flipCard.props.onPress();
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });
});
