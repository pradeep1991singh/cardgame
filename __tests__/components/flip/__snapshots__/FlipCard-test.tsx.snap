// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipCard renders correctly && snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#1ca2f4",
        "borderColor": "#ffffff",
        "borderRadius": 5,
        "borderWidth": 2,
        "height": 150,
        "justifyContent": "center",
        "margin": 5,
        "transform": Array [
          Object {
            "rotateY": "0deg",
          },
        ],
        "width": 100,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontSize": 18,
            "lineHeight": 25,
            "marginVertical": 18,
            "textAlign": "center",
          },
          Object {
            "color": "#ffffff",
          },
        ]
      }
    >
      ?
    </Text>
  </View>
  <View
    collapsable={false}
    style={
      Object {
        "alignItems": "center",
        "backfaceVisibility": "hidden",
        "backgroundColor": "#ffffff",
        "borderColor": "#ffffff",
        "borderRadius": 5,
        "borderWidth": 2,
        "height": 150,
        "justifyContent": "center",
        "margin": 5,
        "position": "absolute",
        "top": 0,
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": 100,
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 18,
          "lineHeight": 25,
          "marginVertical": 18,
          "textAlign": "center",
        }
      }
    >
      1
    </Text>
  </View>
</View>
`;
